src/raster/warp/warp_options.rs

```diff
@@ -0,0 +1,300 @@
+use std::fmt::{Debug, Display, Formatter};
+use std::mem::transmute;
+use std::ptr::NonNull;
+
+use crate::cpl::CslStringList;
+use crate::errors::{GdalError, Result};
+use crate::raster::warp::resample::WarpResampleAlg;
+use crate::raster::GdalDataType;
+use crate::utils::_last_null_pointer_err;
+use crate::xml::GdalXmlNode;
+use gdal_sys::{
+    GDALCloneWarpOptions, GDALCreateWarpOptions, GDALDeserializeWarpOptions,
+    GDALDestroyWarpOptions, GDALSerializeWarpOptions, GDALWarpInitDefaultBandMapping,
+    GDALWarpInitDstNoDataReal, GDALWarpInitSrcNoDataReal, GDALWarpOptions,
+    GDALWarpResolveWorkingDataType,
+};
+use libc::c_char;
+
+/// Container for options provided to GDAL Warp routines.
+///
+/// See: [`GDALWarpOptions`](https://gdal.org/api/gdalwarp_cpp.html#_CPPv415GDALWarpOptions)
+/// for details.
+pub struct GdalWarpOptions(NonNull<GDALWarpOptions>);
+
+impl GdalWarpOptions {
+    pub fn new() -> Self {
+        unsafe { Self::from_ptr(GDALCreateWarpOptions()) }
+    }
+
+    /// Create Self from a raw pointer.
+    ///
+    /// # Safety
+    /// Caller is responsible for ensuring `ptr` is not null, and
+    /// ownership of `ptr` is properly transferred
+    pub unsafe fn from_ptr(ptr: *mut GDALWarpOptions) -> Self {
+        Self(NonNull::new_unchecked(ptr))
+    }
+
+    /// Specify the resampling algorithm to use in Warp operation.
+    pub fn with_resampling_alg(&mut self, alg: WarpResampleAlg) -> &mut Self {
+        unsafe { (*self.as_ptr_mut()).eResampleAlg = alg.to_gdal() };
+        self
+    }
+
+    /// Get the resampling algorithm to be used in Warp operation.
+    pub fn resampling_alg(&self) -> WarpResampleAlg {
+        // `unwrap` below is ok because `with_resampling_alg` is the only way it got set,
+        // aside from the GDAL default, which is `GRA_NearestNeighbour`.
+        WarpResampleAlg::from_gdal(unsafe { (*self.as_ptr()).eResampleAlg }).unwrap_or_default()
+    }
+
+    /// Set the datatype used during processing.
+    ///
+    /// If unset, the algorithm picks the datatype.
+    pub fn with_working_datatype(&mut self, dt: GdalDataType) -> &mut Self {
```

Comment:
Nit: do we use `datatype` or `data_type`?

Reply:
We're not consistent by any stretch:

```
❯ rg --count-matches --ignore-case datatype --glob '*.rs' src
src/driver.rs:12
src/errors.rs:2
src/raster/tests.rs:6
src/raster/types.rs:177
src/raster/mod.rs:4
src/raster/rasterize.rs:1
src/raster/rasterband.rs:9
src/raster/mdarray.rs:76
```

```
❯ rg --count-matches --ignore-case 'data[-_]type' --glob '*.rs' src
src/raster/mdarray.rs:25
src/errors.rs:2
src/driver.rs:2
src/raster/types.rs:3
```

Reply:
I have a bias toward the one word "datatype" form. 

Since Wiktionary is our authoritative source for justifying our own spelling biases, I think I'm justified! 🤣 

https://en.wiktionary.org/wiki/datatype